### PR TITLE
Prevent ALTERing non-existing KS with tablets

### PR DESCRIPTION
**Pull request by ptrsmrn**
ALTER tablets KS executes in 2 steps:
1. ALTER KS's cql handler forms a global topo req, and saves data required to execute this req,
2. global topo req is executed by topo coordinator, which reads data attached to the req.

The KS name is among the data attached to the req. There's a time window between these steps where a to-be-altered KS could have been DROPped, which results in topo coordinator forever trying to ALTER a non-existing KS. In order to avoid it, the code has been changed to first check if a to-be-altered KS exists, and if it's not the case, it doesn't perform any schema/tablets mutations, but just removes the global topo req from the coordinator's queue.
BTW. just adding this extra check resulted in broader than expected changes, which is due to the fact that the code is written badly and needs to be refactored - an effort that's already planned under #19126
(I suggest to disable displaying whitespace differences when reviewing this PR).

Fixes: #19576

Requires 6.0 backport

### CODE REVIEW

**Comment by ptrsmrn** [comment]
@kbr-scylla I was thinking about adding a UT for this change, something along the lines of:
```
# test/cql-pytest/test_keyspace.py
def test_alter_dropped_tablets_keyspace(cql, this_dc, skip_without_tablets):
    keyspace = "to_be_altered_keyspace"
    cql.execute(f"CREATE KEYSPACE {keyspace} WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', '{this_dc}' : 1 }} AND TABLETS = {{'enabled': true}}")
    cql.execute_async(f"ALTER KEYSPACE {keyspace} WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', '{this_dc}': 1}}")
    cql.execute(f'DROP KEYSPACE IF EXISTS {keyspace}')
```
which, theoretically, would create a time window allowing to drop a KS that hasn't been ALTERed by the topo coordinator yet, which is achieved by scheduling ALTER call asynchronously (not waiting for it to complete), so that the following DROP statement could interfere with the above's ALTER. But executing this test 10k times didn't reproduce the issue seen in #19576, so it's not that easy to reproduce it + I think that we already have a testcase that is able to test this scenario - exactly the same test that discovered it (#19576).

**Comment by kbr-scylla** [comment]
@ptrsmrn relying on longevity tests for regression testing is not a good idea.

The PR should come with a regression test -- best would be to have one that fails before the fix, and does not fail after.

You should be able to **very easily** orchestrate the scenario of:
- start ALTER, it commits the request for coordinator to pick up
- perform DROP
- coordinator picks up the request

Using our error injection framework. I don't know if this framework is accessible from cql-pytest, if not, just write a topology_custom test (even single node test should suffice IIUC).

You can inject a wait into topology coordinator at the beginning of handling the request. This injection will wait for a message from the test. Then the test will perform a DROP. Then it will message the injection so coordinator unblocks and handles the request.
See:
-  `future<> inject(const std::string_view& name, waiting_handler_fun func, bool share_messages)` in utils/error_injection.hh
- test/topology_custom/test_raft_no_quorum.py: `test_quorum_lost_during_node_join` for example usage. In particular how `injection_handler` is used

File: service/topology_coordinator.cc [hunk 1]
```diff
@@ -773,35 +771,41 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
...
-                        auto new_strategy = locator::abstract_replication_strategy::create_replication_strategy("NetworkTopologyStrategy", params);
-                        new_tablet_map = co_await new_strategy->maybe_as_tablet_aware()->reallocate_tablets(table, tmptr, old_tablets);
-                    } catch (const std::exception& e) {
-                        error = e.what();
-                        rtlogger.error("Couldn't process global_topology_request::keyspace_rf_change, error: {},"
-                                       "desired new ks opts: {}", error, new_ks_props.get_replication_options());
-                        updates.clear(); // remove all tablets mutations ...
-                        break;           // ... and only create mutations deleting the global req
-                    }
+                    for (const auto& table : ks.metadata()->tables()) {
```
Comment (kbr-scylla) [inline service/topology_coordinator.cc:780]: Please split indentation changes to a separate commit so the PR is easier to review.

**Review by kbr-scylla — CHANGES_REQUESTED** [review]

File: service/topology_coordinator.cc (hunk 1 above)
Comment (ptrsmrn) [reply service/topology_coordinator.cc:780]: I left a note in the description that touches on whitespaces subject:
> (I suggest to disable displaying whitespace differences when reviewing this PR).

So now, so that I remember it once and for all, are maintainers, you included, good with just disable whitespaces diff or is that I always should extract the "refactor" part into a separate commit?

Comment (kbr-scylla) [reply service/topology_coordinator.cc:780]: Interesting, I didn't know about this feature.

I don't know what to decide and don't have a strong opinion. One potential reason for extracting whitespace changes to separate commits is that it improves experience when browsing the `git log` -- which I sometimes do when e.g. debugging a regression, and trying to find where it happened. In theory I could in that case find the PR and there use the whitespace hiding feature, but that's more jumping and I don't know.

@scylladb/scylla-maint any opinions on this matter?

Comment (ptrsmrn) [reply service/topology_coordinator.cc:780]: I'm not sure about `git log`, but `git blame` has extra options to ignore whitespace changes and detect movements of code within/outside of a file, so refactoring the code, i.e. modifying whitespaces or moving it, shouldn't make it that harder to track down a bug.

Comment (denesb) [reply service/topology_coordinator.cc:780]: But I don't want to disable whitespace changes when reviewing PRs, making sure whitespace is just right is the single most important duty of a maintainer!
On a serious note, I don't think ignoring whitespace while reviewing is a good idea. It is a good idea when using git blame.
So let's keep splitting reformatting into separate patches, as before.

Comment (xemul) [reply service/topology_coordinator.cc:780]: I second Botond's suggestion -- though toggling display of whitespaces differences is possible, it doesn't make review easier

Comment (ptrsmrn) [reply service/topology_coordinator.cc:780]: Thanks, I split the changes.

**Comment by ptrsmrn** [comment]
@kbr-scylla thanks for the feedback, suggestions applied, please re-review.

File: service/topology_coordinator.cc [hunk 2]
```diff
@@ -803,6 +804,9 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                         co_await coroutine::maybe_yield();
                     });
                 }
+                } else {
+                    error = "Can't ALTER keyspace " + ks_name + ", keyspace doesn't exist";
+                }
 
                 updates.push_back(canonical_mutation(topology_mutation_builder(guard.write_timestamp())
```
Comment (kbr-scylla) [inline service/topology_coordinator.cc:811]: Shouldn't we skip entering `tablet_migration` and increasing topology version if ALTER fails (there's an `error`)?

Comment (kbr-scylla) [reply service/topology_coordinator.cc:811]: I see that this problem is pre-existing, but I'd like to understand why is this done so nevertheless, and maybe it needs a separate issue.

File: service/topology_coordinator.cc [hunk 3]
```diff
@@ -2898,6 +2898,12 @@ future<> topology_coordinator::run() {
                 co_await await_event();
                 rtlogger.debug("topology coordinator fiber got an event");
             }
+            co_await utils::get_local_injector().inject("wait-before-handling-keyspace-rf-change", [] (auto& handler) -> future<> {
```
Comment (kbr-scylla) [inline service/topology_coordinator.cc:2901]: Let's name this injection more generically. It has nothing to do with keyspace RF change -- it's a general injection to sleep after getting event in topology coordinator fiber. Currently it's only used in test for keyspace rf change, but we may want to reuse it in other tests in other contexts in the future.

File: test/topology_custom/test_tablets.py [hunk 4]
```diff
@@ -301,3 +301,51 @@ async def test_read_of_pending_replica_during_migration(manager: ManagerClient,
 
     rows = await cql.run_async("SELECT pk from test.test")
     assert len(list(rows)) == 1
+
+
+@pytest.mark.asyncio
+@skip_mode('release', 'error injections are not supported in release mode')
+async def test_alter_dropped_tablets_keyspace(manager: ManagerClient) -> None:
```
Comment (kbr-scylla) [inline test/topology_custom/test_tablets.py:308]: Let's put this in separate file so the test case can run concurrently with other test cases in test_tablets.py.

Also, does this test fail without the fix?

File: test/topology_custom/test_tablets.py [hunk 5]
```diff
@@ -301,3 +301,51 @@ async def test_read_of_pending_replica_during_migration(manager: ManagerClient,
...
+                                      "tablets = {'enabled': true}")
+    await manager.get_cql().run_async("create table ks.t (pk int primary key)")
+
+    logger.info(f"injecting wait-before-handling-keyspace-rf-change into the leader node {servers[0]}")
+    injection_handler = await inject_error_one_shot(manager.api, servers[0].ip_addr,
+                                                    'wait-before-handling-keyspace-rf-change')
+
+    async def alter_tablets_ks_without_waiting_to_complete():
+        res = await manager.get_cql().run_async("select data_center from system.local")
+        this_dc = res[0].data_center  # ALTER tablets KS needs to have the DC specified explicitly
```
Comment (kbr-scylla) [inline test/topology_custom/test_tablets.py:330]: Why do you need to get DC from CQL though? Note that the node you get DC from, and the one you execute the statement later, could be different, so if they booted in different DCs, this code would not make much sense. Well it's not a problem in this test because they should boot in the same DC. I guess the only thing you want to do here is to ensure you pass a DC that exists? (and don't care which DC in particular it is) You could make it more obvious in the comments.

File: test/topology_custom/test_tablets.py [hunk 6]
```diff
@@ -301,3 +301,51 @@ async def test_read_of_pending_replica_during_migration(manager: ManagerClient,
...
+        manager.get_cql().run_async("alter keyspace ks "
+                                    f"with replication = {{'class': 'NetworkTopologyStrategy', '{this_dc}': 1}}")
```
Comment (kbr-scylla) [inline test/topology_custom/test_tablets.py:332]: We should `await` for it and verify that it returns an error result at the end of the test.

You decided to spawn a task (`asyncio.create_task`) so the test case an progress concurrently while this is happening.

At the end of the test we can `await` the returned task:
```
task = asyncio.create_task(alter_tablets...)

... do the other stuff, drop keyspace, message handler ...

try:
    await task
except:
   ... verify that the returned exception is as expected (failed ALTER) ...
else:
   fail the test: we expected an exception, not successful response
```

File: test/topology_custom/test_tablets.py [hunk 7]
```diff
@@ -301,3 +301,51 @@ async def test_read_of_pending_replica_during_migration(manager: ManagerClient,
...
+
+    # by creating a task this way we ensure it's immediately executed, but we won't wait until it's completed
+    await asyncio.create_task(alter_tablets_ks_without_waiting_to_complete())
```
Comment (kbr-scylla) [inline test/topology_custom/test_tablets.py:335]: No point to `create_task` and then await it immediately in the foreground... This is equivalent to `await alter_tablets_ks_without_waiting_to_complete()`. But see my comments above. We should have a task but keep it in background and await at the end to verify the result.

**Review by kbr-scylla — CHANGES_REQUESTED** [review]

File: service/topology_coordinator.cc (hunk 2 above)
Comment (ptrsmrn) [reply service/topology_coordinator.cc:811]: If preparing ALTER mutations (schema mutations or tablets mutations) fails, we still need to delete the global `rf_change` req along with the topo req tracking entry. 
So whether ALTER succeeds or not, we need to perform some mutations.

File: test/topology_custom/test_tablets.py (hunk 4 above)
Comment (ptrsmrn) [reply test/topology_custom/test_tablets.py:308]: Yes, the test fails without the fix.

File: test/topology_custom/test_tablets.py (hunk 5 above)
Comment (ptrsmrn) [reply test/topology_custom/test_tablets.py:330]: I need to explicitly state the name of the DC, because ALTER tablets KS doesn't accept the generic `replication_factor` tag.

File: test/topology_custom/test_tablets.py (hunk 4 above)
Comment (ptrsmrn) [reply test/topology_custom/test_tablets.py:308]: What's the reason for moving this specific testcase to a separate file? You want to test if ALTER can indeed run concurrently with other operations? Testing from this angle sounds OK, but this approach may not be the best, understanding why a testcase failed would be hard.

File: service/topology_coordinator.cc (hunk 3 above)
Comment (ptrsmrn) [reply service/topology_coordinator.cc]: Right, changed to `wait-after-topology-coordinator-gets-event`.

File: test/topology_custom/test_tablets.py (hunk 5 above)
Comment (ptrsmrn) [reply test/topology_custom/test_tablets.py]: Changed the comment to be more specific: `# ALTER tablets KS only accepts a specific DC, it rejects the generic 'replication_factor' tag`

File: test/topology_custom/test_tablets.py (hunk 6 above)
Comment (ptrsmrn) [reply test/topology_custom/test_tablets.py:333]: Rewritten, thanks.

File: test/topology_custom/test_tablets.py (hunk 7 above)
Comment (ptrsmrn) [reply test/topology_custom/test_tablets.py]: Addressed in https://github.com/scylladb/scylladb/pull/19666#discussion_r1707772916

**Comment by ptrsmrn** [comment]
@kbr-scylla can you please re-eview?

File: service/topology_coordinator.cc (hunk 2 above)
Comment (kbr-scylla) [reply service/topology_coordinator.cc:811]: This doesn't answer my question. Yes we need to perform **some** mutations, the ones you mentioned. But what about these things that I asked:
> entering tablet_migration and increasing topology version

?

File: test/topology_custom/test_tablets.py (hunk 4 above)
Comment (kbr-scylla) [reply test/topology_custom/test_tablets.py:308]: > What's the reason for moving this specific testcase to a separate file?

Speed of the test suite i.e. how fast test suite can finish (the more test cases can run concurrently, the faster the whole suite will finish)

> You want to test if ALTER can indeed run concurrently with other operations?

No, each test case has exclusive access to a leased cluster...

**Comment by kbr-scylla** [comment]
LGTM except:
- https://github.com/scylladb/scylladb/pull/19666#discussion_r1715208314
- https://github.com/scylladb/scylladb/pull/19666#discussion_r1715206430 (pre-existing problem, but I think an issue needs to be created)

File: service/topology_coordinator.cc (hunk 2 above)
Comment (ptrsmrn) [reply service/topology_coordinator.cc:811]: That needs to be investigated, I don't know the answer off the top of my head. Added to https://github.com/scylladb/scylladb/issues/19123

File: test/topology_custom/test_tablets.py (hunk 4 above)
Comment (ptrsmrn) [reply test/topology_custom/test_tablets.py]: Clear, thanks, done.

**Comment by ptrsmrn** [comment]
> LGTM except:
> 
> * [Prevent ALTERing non-existing KS with tablets #19666 (comment)](https://github.com/scylladb/scylladb/pull/19666#discussion_r1715208314)
> * [Prevent ALTERing non-existing KS with tablets #19666 (comment)](https://github.com/scylladb/scylladb/pull/19666#discussion_r1715206430) (pre-existing problem, but I think an issue needs to be created)

@kbr-scylla addressed the comments, the CI is green.

**Review by kbr-scylla — APPROVED** [review]
